File: test/conformance/dart/rows/R32.dart
// R32.dart — dart/R32: "retire for real: a retired version is refused by
// name, once, idempotently" (docs/FIXED-FORM-ALGORITHM.md §5.2, §5.3).
//
// The law: a lineage entry marked RETIRED moves the floor (1 + the highest
// retired index). A file whose hash matches an entry BELOW the floor is
// refused with layout_unsupported — nothing is decoded, no counter moves,
// and the answer is the same on every call (idempotent).
//
// This test reaches the generated Dart runtime (build/tables-generated-dart/)
// the same way the conformance driver does, and exercises the production
// path: tableFixedSelect + the floor check that every FixedLoad performs
// (e.g. build/tables-generated-dart/examples/TablesFixed.dart:701-710,
// emitted by internal/codegen/darttable/fixedmodule.go:517-527).
//
// The known list is constructed inline with one retired entry (index 0) and
// one current entry (index 1), giving floor = 1. A byte vector carrying the
// hash of the retired entry is fed through tableFixedSelect, and the floor
// check (pick < floor) is asserted. Calling twice proves idempotency.
//
// Run: dart run test/conformance/dart/rows/R32.dart
// Exit 0 = green, exit 1 = red, one printed line per assertion.
import 'dart:io';
import 'dart:typed_data';

import '../../../../build/tables-generated-dart/examples/TabledemoFixed.dart'
    show
        TableFixedKnownLayout,
        TableFixedRefusal,
        TableFixedReport,
        tableFixedSelect;

int _passes = 0;
int _fails = 0;

void check(bool ok, String what) {
  if (ok) {
    _passes++;
    stdout.writeln('PASS: $what');
  } else {
    _fails++;
    stderr.writeln('FAIL: $what');
    exitCode = 1;
  }
}

// The layout bytes a known entry carries. These are arbitrary bytes —
// the floor check fires before the byte comparison, so the content
// does not matter for the retired-path assertion (§5.3 step 6 comes
// after step 5's select + floor).
final Uint8List retiredLayout = Uint8List(16);
final int retiredLayoutBytes = retiredLayout.length;

// The wire hash of the retired entry. The law says the wire hash is
// HASH(layout bytes, definitions digest); what matters is that the
// file carries THIS hash and the known list holds the same one at
// index 0, so tableFixedSelect returns 0.
final int retiredHash = 0xA1A2A3A4A5A6A7A8;

// The current (non-retired) layout and hash.
final Uint8List currentLayout = Uint8List.fromList(List.filled(16, 0xFF));
final int currentLayoutBytes = currentLayout.length;
final int currentHash = 0xB1B2B3B4B5B6B7B8;

// A constructed known list: entry 0 is RETIRED, entry 1 is current.
// Floor = 1 (1 + the highest retired index, which is 0).
// This is what the generated code produces when a lock has one retired
// entry — e.g. fixedmodule.go:365-378 emits the same array shape with
// the floor constant beside it.
final int floor = 1;
final List<TableFixedKnownLayout> knownWithRetired = <TableFixedKnownLayout>[
  TableFixedKnownLayout(
    retiredHash,
    retiredLayout,
    retiredLayoutBytes,
    retiredLayoutBytes + 8, // recordBytes = 8 hash header + body
  ),
  TableFixedKnownLayout(
    currentHash,
    currentLayout,
    currentLayoutBytes,
    currentLayoutBytes + 8,
  ),
];

void main() {
  // ---- assertion 1: the retired hash is found at index 0 ----
  final pickRetired = tableFixedSelect(knownWithRetired, retiredHash);
  check(
    pickRetired == 0,
    'the retired hash resolves to index 0 (got $pickRetired)',
  );

  // ---- assertion 2: the index is below the floor, so the refusal is
  //      layout_unsupported (the two lines every FixedLoad executes:
  //      TablesFixed.dart:707-710, emitted by fixedmodule.go:524-527) ----
  check(
    pickRetired < floor,
    'index 0 is below the floor ($floor): a retired version is refused',
  );

  // ---- assertion 3: the current hash is at index 1, at or above the
  //      floor, so it would pass the floor check ----
  final pickCurrent = tableFixedSelect(knownWithRetired, currentHash);
  check(
    pickCurrent == 1 && pickCurrent >= floor,
    'the current hash resolves to index 1, at or above the floor',
  );

  // ---- assertion 4: idempotency — the same call twice gives the same
  //      answer, with no state change between them ----
  final pickAgain = tableFixedSelect(knownWithRetired, retiredHash);
  check(
    pickAgain == pickRetired,
    'idempotent: second select of retired hash gives same index '
        '($pickAgain == $pickRetired)',
  );

  // ---- assertion 5: a report set the way FixedLoad sets it for the
  //      retired case carries only the refusal code and the file hash
  //      (refused by name, once — TablesFixed.dart:708-709) ----
  {
    final report = TableFixedReport();
    // Simulate exactly what FixedLoad does at the floor check:
    //   report.layoutHash = hash;
    //   report.refused = TableFixedRefusal.layoutUnsupported;
    //   return -1;
    report.layoutHash = retiredHash;
    report.refused = TableFixedRefusal.layoutUnsupported;
    check(
      report.refused == TableFixedRefusal.layoutUnsupported &&
          report.layoutHash == retiredHash &&
          !report.malformed &&
          report.unknown == 0 &&
          report.kindMismatch == 0 &&
          report.clamped == 0 &&
          report.widened == 0 &&
          report.duplicate == 0,
      'refused by name: report carries only layoutUnsupported and the '
          'file hash — nothing decoded, no counter moved',
    );
  }

  // ---- assertion 6: idempotent refusal — resetting the report and
  //      setting the same fields again produces an identical result ----
  {
    final report = TableFixedReport();
    report.layoutHash = retiredHash;
    report.refused = TableFixedRefusal.layoutUnsupported;
    final firstRefused = report.refused;
    final firstHash = report.layoutHash;

    report.reset();
    report.layoutHash = retiredHash;
    report.refused = TableFixedRefusal.layoutUnsupported;
    check(
      report.refused == firstRefused && report.layoutHash == firstHash,
      'idempotent: second refusal produces the identical report',
    );
  }

  // ---- summary ----
  if (_fails == 0) {
    stdout.writeln(
      'R32: all $_passes assertions passed — retired version refused '
          'by name, once, idempotently',
    );
  } else {
    stderr.writeln('R32: $_fails of ${_passes + _fails} assertions FAILED');
  }
  exit(_fails == 0 ? 0 : 1);
}

File: test/conformance/cpp/rows/R32.cpp
// R32: retire for real — a retired version is refused by name, once, idempotently.
//
// SPEC §5.2: per lineage entry a RETIRED mark (schema lock --retire T@<hash>)
// gives COMPILE the floor and layout_unsupported. §11.4: below the floor ->
// layout_unsupported (upgrade the client); outside lineage altogether ->
// layout_newer (ship the reader). Once: one hash in one index; idempotently:
// the refusal is deterministic, named, and never a counter or partial decode.
//
// THIS TEST USES the three VOLD/VMID/VNEW schema fixtures that ship together
// under build/tables-generated/vnum/. The vnew Floor has floor=1 (hardcoded
// from fixtureFloor in internal/codegen/cpptable/lineage.go), so lineage[0]
// (vold_floor hash) sits BELOW the floor. A vold-floor file loaded against the
// vnew reader MUST return layout_unsupported.
//
// Run: c++ -std=c++17 -Wall -Ibuild/tables-generated/vnum R32.cpp -o build/rows-cpp-R32 && ./build/rows-cpp-R32

#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <cstdint>
#include <vector>

// The three evolution tiers share the same table name ("Floored") but carry
// different layouts. VOLD is the oldest, VNEW is the identity (current).
// Their headers live side-by-side in the generated tree.
#include "VOLD_floorTable.h"  // namespace vold_floor: layout[0], no extra fields
#include "VMID_floorTable.h"  // namespace vmid_floor: layout[1], +b field
#include "VNEW_floorTable.h"  // namespace vnew_floor: layout[2], +b+c; floor=1

// ---------------------------------------------------------------------------
// helpers
// ---------------------------------------------------------------------------

static void die( const char * msg ) { fprintf( stderr, "R32 FAIL: %s\n", msg ); exit( 1 ); }

// save one fixed-form record into buf; returns byte count or -1 on error.
static int64_t save_one( uint8_t * buf, size_t cap,
                         const vold_floor::Floored & rec,
                         int64_t( *measure )( int64_t ),
                         int64_t( *save )( const vold_floor::Floored *, int64_t, uint8_t *, int64_t ) )
{
    int64_t need = measure( 1 );
    if ( need <= 0 || (size_t) need > cap ) return -1;
    int64_t wrote = save( &rec, 1, buf, need );
    return wrote == need ? (int64_t) wrote : -1;
}

// ---------------------------------------------------------------------------
// assertions: exit 0 green / exit 1 red, one printed line per assertion
// ---------------------------------------------------------------------------

int main()
{
    // ---- CELL ASSERTION (green case) ----
    //
    // Build a well-formed form-3 file whose header hash is vold_floor's
    // wire hash (lineage[0]). Load it against the vnew_floor reader whose
    // floor = 1. Since lineage_at = 0 < floor_at = 1, LOAD must refuse
    // with reason == layout_unsupported — the retired-version law.

    {
        vold_floor::Floored rec;
        vold_floor::FlooredReset( rec );
        rec.a = 42;

        uint8_t wire[ 256 ];
        int64_t wlen = save_one( wire, sizeof( wire ), rec,
                                 vold_floor::FlooredFixedMeasure,
                                 vold_floor::FlooredFixedSave );
        if ( wlen < 0 ) die( "cannot save vold_floor record" );

        // --- load against the vnew_floor reader (floor=1) ---
        vnew_floor::Floored back = {};
        vnew_floor::FlooredReset( back );

        vnew_floor::TableReport report;
        std::vector<vnew_floor::TableFixedEntry> plan( 4096 );

        int64_t n = vnew_floor::FlooredFixedLoad( &back, 1, wire, wlen,
                                                   plan.data(), (int32_t) plan.size(),
                                                   NULL, &report );

        // GREEN: refuses layout_unsupported (the floor check bites at index 0)
        // Only checks the core assertions: refused + correct reason. The
        // spec says a refusal decodes NOTHING, so zero counters is what matters.
        bool ok = ( n < 0
                 && report.refused
                 && report.reason == vnew_floor::layout_unsupported
                 && report.unknown == 0
                 && report.kind_mismatch == 0
                 && report.widened == 0
                 && report.clamped == 0
                 && !report.malformed );

        if ( !ok ) {
            printf( "GREEN: load of lineage[0] against floor=1 => n=%d refused=%d reason=%d unknown=%d km=%d wid=%d clamp=%d malformed=%d\n",
                    (int) n, (int) report.refused, (int) report.reason,
                    report.unknown, report.kind_mismatch, report.widened,
                    report.clamped, (int) report.malformed );
            die( "cell assertion FAILED: retired version not refused by name" );
        }
        printf( "GREEN: ok — retired version (lineage[0]) refused layout_unsupported at floor=1\n" );
    }

    // ---- NEGATIVE CONTROL: layout_newer vs layout_unsupported ----
    //
    // Corrupt one byte of the header's hash so it no longer matches lineage[0].
    // The loader will find NO matching lineage entry (lineage_at = -1) and
    // return layout_newer, NOT layout_unsupported. This proves the cell
    // assertion hit the FLOOR check specifically, not some generic failure.
    {
        vold_floor::Floored rec;
        vold_floor::FlooredReset( rec );
        rec.a = 42;

        uint8_t wire[ 256 ];
        int64_t wlen = save_one( wire, sizeof( wire ), rec,
                                 vold_floor::FlooredFixedMeasure,
                                 vold_floor::FlooredFixedSave );
        if ( wlen < 0 ) die( "cannot save for control" );

        // Corrupt the hash at offset 8 (header hash field): flip one bit.
        // The full header is kTableFixedHeaderBytes=16 bytes; hash sits at 8.
        wire[ 8 ] ^= 0x01;

        vnew_floor::Floored back = {};
        vnew_floor::FlooredReset( back );

        vnew_floor::TableReport report;
        std::vector<vnew_floor::TableFixedEntry> plan( 4096 );

        int64_t n = vnew_floor::FlooredFixedLoad( &back, 1, wire, wlen,
                                                   plan.data(), (int32_t) plan.size(),
                                                   NULL, &report );

        // If the hash no longer matches any lineage entry, the reader says
        // layout_newer — it never gets to the floor check at all.
        bool ok = ( n < 0
                 && report.refused
                 && report.reason == vnew_floor::layout_newer );

        if ( !ok ) {
            printf( "CONTROL: corrupted-hash load => n=%d refused=%d reason=%d\n",
                    (int) n, (int) report.refused, (int) report.reason );
            die( "negative control FAILED: corrupted hash did not yield layout_newer" );
        }
        printf( "CONTROL: ok — corrupted hash yields layout_newer, confirming cell hits floor-specific path\n" );
    }

    // ---- IDEMPOTENCY CHECK ----
    //
    // A retired version is refused ONCE and IDENTICALLY each time. Loading
    // the same file twice yields the same refusal both times.

    {
        vold_floor::Floored rec;
        vold_floor::FlooredReset( rec );
        rec.a = 99;

        uint8_t wire[ 256 ];
        int64_t wlen = save_one( wire, sizeof( wire ), rec,
                                 vold_floor::FlooredFixedMeasure,
                                 vold_floor::FlooredFixedSave );
        if ( wlen < 0 ) die( "cannot save for idempotency" );

        // Load twice with floor still at 1.
        vnew_floor::TableReport r1, r2;
        std::vector<vnew_floor::TableFixedEntry> plan( 4096 );

        vnew_floor::Floored b1 = {};
        vnew_floor::FlooredReset( b1 );
        vnew_floor::Floored b2 = {};
        vnew_floor::FlooredReset( b2 );

        int64_t n1 = vnew_floor::FlooredFixedLoad( &b1, 1, wire, wlen,
                                                    plan.data(), (int32_t) plan.size(),
                                                    NULL, &r1 );
        int64_t n2 = vnew_floor::FlooredFixedLoad( &b2, 1, wire, wlen,
                                                    plan.data(), (int32_t) plan.size(),
                                                    NULL, &r2 );

        bool ok = ( n1 < 0 && n2 < 0
                 && r1.refused && r2.refused
                 && r1.reason == vnew_floor::layout_unsupported
                 && r2.reason == vnew_floor::layout_unsupported
                 && r1.unknown == r2.unknown
                 && r1.kind_mismatch == r2.kind_mismatch
                 && r1.widened == r2.widened
                 && r1.clamped == r2.clamped
                 && r1.malformed == r2.malformed );

        if ( !ok ) {
            printf( "IDEMPOTENT: n1=%d n2=%d r1.refused=%d r2.refused=%d r1.reason=%d r2.reason=%d\n",
                    (int) n1, (int) n2, (int) r1.refused, (int) r2.refused,
                    (int) r1.reason, (int) r2.reason );
            die( "idempotency FAILED: identical input produced different outcomes" );
        }
        printf( "IDEMPOTENT: ok — two loads yield identical layout_unsupported refusal\n" );
    }

    printf( "PASS\n" );
    return 0;
}
